lib/libc/gen/sig2str.c

```diff
@@ -0,0 +1,104 @@
+/*-
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright (c) 2025 Ricardo Branco <rbranco@suse.de>.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *      notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *      notice, this list of conditions and the following disclaimer in the
+ *      documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+/*
+ * Translate between signal names and numbers
+ */
+#include "namespace.h"
+#include <ctype.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "un-namespace.h"
+
+static const char rtmin_str[] = "RTMIN";
+static const char rtmax_str[] = "RTMAX";
+
+int
+sig2str(int signum, char *str)
+{
+	if (signum <= 0 || signum > SIGRTMAX)
+		return (-1);
+
+	if (signum < sys_nsig)
+		(void)strcpy(str, sys_signame[signum]);
+	else if (signum < SIGRTMIN)
+		(void)sprintf(str, "%d", signum);
+	else if (signum == SIGRTMIN)
+		(void)strcpy(str, rtmin_str);
+	else if (signum == SIGRTMAX)
+		(void)strcpy(str, rtmax_str);
+	else if (signum <= (SIGRTMIN + SIGRTMAX) / 2)
+		(void)sprintf(str, "%s+%d", rtmin_str, signum - SIGRTMIN);
+	else
+		(void)sprintf(str, "%s-%d", rtmax_str, SIGRTMAX - signum);
+
+	return (0);
+}
+
+int
+str2sig(const char * restrict str, int * restrict pnum)
+{
+	int n, sig;
+	char *end;
+	int rtend = sizeof(rtmin_str) - 1;
+
+	if (strncmp(str, rtmin_str, sizeof(rtmin_str) - 1) == 0 ||
+	    strncmp(str, rtmax_str, sizeof(rtmax_str) - 1) == 0) {
+		sig = (str[4] == 'X') ? SIGRTMAX : SIGRTMIN;
+		n = 0;
+		if (str[rtend] == '+' || str[rtend] == '-') {
+			n = (int) strtol(str + rtend, &end, 10);
+			if (*end != '\0' || n == 0)
+				return (-1);
+		} else if (str[rtend] != '\0')
+			return (-1);
+		sig += n;
+		if (sig < SIGRTMIN || sig > SIGRTMAX)
+			return (-1);
+		*pnum = sig;
+		return (0);
+	}
+
+	if (isdigit(str[0])) {
+		sig = (int)strtol(str, &end, 10);
```

Comment:
Please use an appropriately-sized variable instead of a cast, and check for overflow (see `strtol(3)` for details).